scripts/dev/tidy.php

```diff
@@ -0,0 +1,571 @@
+
+  SYNOPSIS:
+
+    php tidy.php [OPTION...] input-path
+
+  DESCRIPTION:
+
+    This script checks and fixes common whitespace issues. It can trim
+    trailing whitespace (spaces and tabs) from each line, insert missing
+    final newline, and trim multiple final newlines into a single newline.
+    If file has multiple different EOL (end of line) characters - LF (\\n),
+    CRLF (\\r\\n) or CR (\\r), the prevailing EOL character is used as a
+    final newline.
+
+    When used on the php-src root directory it ignores the predefined set
+    of binary files and certain test files, and works only on whitelisted
+    set of predefined files for convenience.
+
+    By default, original files are backed up (file.ext~).
+
+  OPTIONS:
+
+    -h, --help       Display this help
+
+    -w, --trim-trailing-whitespace
+                    Trim trailing whitespace (spaces and tabs)
+
+    -n, --insert-final-newline
+                     Insert a final EOL character if one is missing.
+                     Redundant final newlines or final spaces are not trimmed.
+
+    -N, --trim-final-newlines [NUM]
+                     Fix end of file (trim redundant final newlines and
+                     insert a single final newline using a prevailing EOL
+                     character (LF or CRLF) in current file. If NUM is set,
+                     allow NUM of redundant trailing final newlines.
+                     Default NUM=1
+
+                     According to POSIX, a line is a sequence of zero or
+                     more non-' <newline>' characters plus a terminating
+                     '<newline>' character.
+
+        --no-colors  Disable colors in the script output
+
+        --no-backup  Disable backed up files. Use with caution to not
+                     overwrite something by accident.
+
+    -f, --fix        Edit and save files in place
+
+    -q, --quiet      Do not output any message
+
+  EXAMPLES:
+
+    Display help: php tidy.php -h
+
+    Check php-src: scripts/dev/tidy.php .
+
+    Fix php-src: scripts/dev/tidy.php --fix --no-backup .
+
+    Fix custom path or file: ./tidy.php --fix input-path
+
+    Trim only trailing whitespace: ./tidy.php -w --fix input-path
+
+HELP;
+}
+
+/**
+ * Return short invalid usage error.
+ */
+function invalidUsage(string $error): string
+{
+    $output = "tidy.php: $error\n";
+    $output .= "Try 'php tidy.php --help' for more information.\n";
+
+    return $output;
+}
+
+/**
+ * Trim trailing spaces and tabs from each line.
+ */
+function trimTrailingWhitespace(string $content): string
+{
+    return preg_replace('/[\t ]+(\r?$)/m', '$1', $content);
+}
+
+/**
+ * If string doesn't have a final newline append one using a prevailing EOL
+ * character in the current string LF (\n), CRLF (\r\n) or CR (\r).
+ */
+function insertFinalNewline(string $content): string
+{
+    if (
+        substr($content, -2) !== "\r\n"
+        && substr($content, -1) !== "\n"
+        && substr($content, -1) !== "\r"
```

Comment:
Hello, thanks for these review info... Very useful ones. The `\r` character is optional and just in case functionality. Yes, the `\n` is now used on Mac and newere *nix systems. Modern editors also recognize the `\r` ending line as `\r\n`... Otherwise to fix file with possible `\r` EOL it's an option that this script also includes.